File: backend/app/schemas/reliability.py
import uuid
from datetime import datetime

from pydantic import BaseModel


class DurableRunEntry(BaseModel):
    id: uuid.UUID
    agent_id: uuid.UUID | None
    agent_name: str | None
    status: str
    adk_session_id: str | None
    adk_invocation_id: str | None
    error_category: str | None
    error_message: str | None
    invoked_by: str | None
    created_at: datetime
    age_seconds: float
    # True once a 'running' row has sat longer than the staleness threshold
    # the list endpoint was called with — a real in-flight request just
    # hasn't finished yet; a stale one almost certainly means the process
    # that owned it died before it could reach its final status update.
    is_stale: bool


class DurableRunListResponse(BaseModel):
    items: list[DurableRunEntry]
    total: int
    limit: int
    offset: int


class DurableRunResumeResponse(BaseModel):
    id: uuid.UUID
    status: str
    response_text: str | None
    error_message: str | None


class CircuitBreakerEntry(BaseModel):
    key: str
    state: str
    consecutive_failures: int
    cooldown_remaining_seconds: float | None

File: frontend/src/api/reliability.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { api } from './client'

export interface DurableRunEntry {
  id: string
  agent_id: string | null
  agent_name: string | null
  status: string
  adk_session_id: string | null
  adk_invocation_id: string | null
  error_category: string | null
  error_message: string | null
  invoked_by: string | null
  created_at: string
  age_seconds: number
  is_stale: boolean
}

export interface DurableRunListResponse {
  items: DurableRunEntry[]
  total: number
  limit: number
  offset: number
}

export interface DurableRunResumeResponse {
  id: string
  status: string
  response_text: string | null
  error_message: string | null
}

export interface CircuitBreakerEntry {
  key: string
  state: string
  consecutive_failures: number
  cooldown_remaining_seconds: number | null
}

const KEY = ['reliability'] as const

export function useDurableRuns(status: string | undefined, offset: number, limit = 25) {
  return useQuery({
    queryKey: [...KEY, 'runs', status ?? 'all', offset, limit],
    queryFn: () => {
      const params = new URLSearchParams({ limit: String(limit), offset: String(offset) })
      if (status) params.set('status', status)
      return api.get<DurableRunListResponse>(`/reliability/runs?${params.toString()}`)
    },
    refetchInterval: 15_000,
  })
}

export function useCircuitBreakers() {
  return useQuery({
    queryKey: [...KEY, 'circuit-breakers'],
    queryFn: () => api.get<CircuitBreakerEntry[]>('/reliability/circuit-breakers'),
    refetchInterval: 15_000,
  })
}

export function useResumeDurableRun() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (invocationLogId: string) =>
      api.post<DurableRunResumeResponse>(`/reliability/runs/${invocationLogId}/resume`),
    onSuccess: () => qc.invalidateQueries({ queryKey: [...KEY, 'runs'] }),
  })
}
